crates/extensions/c8y_http_proxy/src/actor.rs

```diff
@@ -162,80 +166,148 @@ impl C8YHttpProxyActor {
         Ok(())
     }
 
-    async fn try_get_and_set_internal_id(&mut self) -> Result<(), C8YRestError> {
-        let internal_id = self.try_get_internal_id(None).await?;
-        self.end_point.set_c8y_internal_id(internal_id);
+    async fn get_and_set_internal_id(
+        &mut self,
+        device_type: DeviceType,
+    ) -> Result<(), C8YRestError> {
+        let internal_id = self.try_get_internal_id(device_type.clone()).await?;
+        self.end_point.set_internal_id(device_type, internal_id);
+
         Ok(())
     }
 
-    async fn get_c8y_internal_child_id(
+    async fn try_get_internal_id(
         &mut self,
-        child_device_id: String,
+        device_type: DeviceType,
     ) -> Result<String, C8YRestError> {
-        if let Some(c8y_internal_id) = self.child_devices.get(&child_device_id) {
-            Ok(c8y_internal_id.clone())
-        } else {
-            let c8y_internal_id = self.try_get_internal_id(Some(&child_device_id)).await?;
-            self.child_devices
-                .insert(child_device_id, c8y_internal_id.clone());
-            Ok(c8y_internal_id)
+        let device_id = self.end_point.get_device_id(device_type);
+        let url_get_id: String = self.end_point.get_url_for_internal_id(device_id);
+        if self.end_point.token.is_none() {
+            self.get_fresh_token().await?;
         }
-    }
+        let build_request = |end_point: &C8yEndPoint| -> Result<HttpRequestBuilder, C8YRestError> {
+            Ok(HttpRequestBuilder::get(&url_get_id)
+                .bearer_auth(end_point.token.clone().unwrap_or_default()))
+        };
 
-    async fn try_get_internal_id(
-        &mut self,
-        device_id: Option<&str>,
-    ) -> Result<String, C8YRestError> {
-        let url_get_id = self.end_point.get_url_for_get_id(device_id);
+        let request_builder = build_request(&self.end_point)?;
```

Comment:
As already said, I would not use a closure to build the request to get the internal id of a device. I would simply use the url returned by `get_url_for_internal_id()` and not call `try_request_with_fresh_token()`. Sharing the later for regular requests and for the very specific request to get an internal id makes the code more complex than necessary.

Reply:
done

Reply:
I don't see what has been updated.

Reply:
Removed the closure and used the `URL` returned by the `get_url_for_internal_id()`.